File: libs/js/encoder.js
// IWX encoder/decoder utilities (JSON profile)
import { ERROR_CODES, IwxError } from './errors.js';
import { getSchema } from './schema-loader.js';

function validateValueType(expected, value) {
  if (value === null || value === undefined) return true;
  switch (expected) {
    case 'string':
      return typeof value === 'string';
    case 'number':
      return typeof value === 'number';
    case 'integer':
      return Number.isInteger(value);
    case 'boolean':
      return typeof value === 'boolean';
    case 'date':
      return typeof value === 'string' || value instanceof Date;
    case 'object':
      return typeof value === 'object';
    default:
      return true;
  }
}

function mapFieldValue(field, source, obj) {
  const key = source === 'backend' ? field.backend : field.frontend;
  const value = Object.prototype.hasOwnProperty.call(obj, key) ? obj[key] : null;
  if (field.required && (value === null || value === undefined)) {
    throw new IwxError(ERROR_CODES.MISSING_REQUIRED, `Missing required field ${key}`, { field });
  }
  if (field.type && !validateValueType(field.type, value)) {
    throw new IwxError(ERROR_CODES.TYPE_MISMATCH, `Invalid type for ${key}`, { expected: field.type, value });
  }
  return value;
}

function buildValues(schema, obj, source) {
  const values = [];
  for (const field of schema.fields) {
    values[field.pos] = mapFieldValue(field, source, obj);
  }
  return values;
}

export function iwxEncodePlain(schemaId, sourceObject, options = {}) {
  const { source = 'frontend' } = options;
  const schema = getSchema(schemaId);
  return { _s: schemaId, _v: buildValues(schema, sourceObject, source) };
}

export function iwxDecode(payload, options = {}) {
  const { target = 'frontend' } = options;
  const schema = getSchema(payload._s);
  const values = payload._v || [];
  const obj = {};
  for (const field of schema.fields) {
    const value = values[field.pos] ?? null;
    const key = target === 'backend' ? field.backend : field.frontend;
    obj[key] = value;
  }
  return obj;
}

export function iwxDecodeToFront(payload) {
  return iwxDecode(payload, { target: 'frontend' });
}

export function iwxDecodeToBackend(payload) {
  return iwxDecode(payload, { target: 'backend' });
}

File: libs/python/encoder.py
from .schema_loader import get_schema
from .errors import ERROR_CODES, IWXError


def _validate_value(field: dict, obj: dict, source: str):
    key = field['backend'] if source == 'backend' else field['frontend']
    value = obj.get(key)
    if field.get('required') and value is None:
        raise IWXError(ERROR_CODES['MISSING_REQUIRED'], f'Missing required field {key}', {'field': field})
    ftype = field.get('type')
    if ftype:
        if ftype == 'string' and value is not None and not isinstance(value, str):
            raise IWXError(ERROR_CODES['TYPE_MISMATCH'], f'Expected string for {key}')
        if ftype == 'number' and value is not None and not isinstance(value, (int, float)):
            raise IWXError(ERROR_CODES['TYPE_MISMATCH'], f'Expected number for {key}')
        if ftype == 'integer' and value is not None and not isinstance(value, int):
            raise IWXError(ERROR_CODES['TYPE_MISMATCH'], f'Expected integer for {key}')
        if ftype == 'boolean' and value is not None and not isinstance(value, bool):
            raise IWXError(ERROR_CODES['TYPE_MISMATCH'], f'Expected boolean for {key}')
    return value


def _build_values(schema: dict, obj_front: dict, source: str):
    max_pos = max((field['pos'] for field in schema['fields']), default=-1)
    values = [None] * (max_pos + 1)
    for field in schema['fields']:
        values[field['pos']] = _validate_value(field, obj_front, source)
    return values


def iwx_encode_plain(schema_id: int, obj_front: dict, *, source: str = 'frontend') -> dict:
    """Convert a frontend/backend object to a plain IWX payload {_s,_v}."""
    schema = get_schema(schema_id)
    values = _build_values(schema, obj_front, source)
    return {'_s': schema_id, '_v': values}
